### DISCUSSION TITLE
Silent failure when running `airflow db init`

### DISCUSSION THREAD

**Discussion by ziplokk1** [apache/airflow#18408] [Show and tell]
Hey folks,

I've come across an issue where when attempting to run migrations against a clean database, but with existing dags defined in my dags folder, there's a silent failure and the migration exits with the `airflow` command usage prompt saying "invalid command 'db'". 

I got to digging and the culprit was the [following line](https://github.com/apache/airflow/blob/main/airflow/migrations/versions/cc1e65623dc7_add_max_tries_column_to_task_instance.py#L73).

I'm almost positive that this has to do with an erroneous `.py` file in my dags folder (I have yet to dig deeper into the issue), but this begged the question of why is the migration is reading the dags to migrate existing data in the database in the first place? Shouldn't the migration simply migrate the data regardless of the state of the dag files themselves?

I was able to fix the issue by changing the line in the link above to `dagbag = DagBag(read_dags_from_db=True)` but I'd like to see if there was any context behind reading the dag file instead of the database before opening a PR.

**Comment by BasPH** [comment]
Not sure if I understand the issue correctly, but the error "invalid command 'db'" suggests to me you're running `airflow db init` on an Airflow 1 installation (where it was `airflow dbinit`). Could you check your Airflow version?

**Reply by ziplokk1** [reply]
I'm running airflow 2.0.2.

```
❯ airflow version
2.0.2
❯ airflow db init
DB: postgresql+psycopg2://airflow:***@localhost:5432/airflow
[2021-09-22 10:47:02,728] {db.py:684} INFO - Creating tables
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> e3a246e0dc1, current schema
INFO  [alembic.runtime.migration] Running upgrade e3a246e0dc1 -> 1507a7289a2f, create is_encrypted
INFO  [alembic.runtime.migration] Running upgrade 1507a7289a2f -> 13eb55f81627, maintain history for compatibility with earlier migrations
INFO  [alembic.runtime.migration] Running upgrade 13eb55f81627 -> 338e90f54d61, More logging into task_instance
INFO  [alembic.runtime.migration] Running upgrade 338e90f54d61 -> 52d714495f0, job_id indices
INFO  [alembic.runtime.migration] Running upgrade 52d714495f0 -> 502898887f84, Adding extra to Log
INFO  [alembic.runtime.migration] Running upgrade 502898887f84 -> 1b38cef5b76e, add dagrun
INFO  [alembic.runtime.migration] Running upgrade 1b38cef5b76e -> 2e541a1dcfed, task_duration
INFO  [alembic.runtime.migration] Running upgrade 2e541a1dcfed -> 40e67319e3a9, dagrun_config
INFO  [alembic.runtime.migration] Running upgrade 40e67319e3a9 -> 561833c1c74b, add password column to user
INFO  [alembic.runtime.migration] Running upgrade 561833c1c74b -> 4446e08588, dagrun start end
INFO  [alembic.runtime.migration] Running upgrade 4446e08588 -> bbc73705a13e, Add notification_sent column to sla_miss
INFO  [alembic.runtime.migration] Running upgrade bbc73705a13e -> bba5a7cfc896, Add a column to track the encryption state of the 'Extra' field in connection
INFO  [alembic.runtime.migration] Running upgrade bba5a7cfc896 -> 1968acfc09e3, add is_encrypted column to variable table
INFO  [alembic.runtime.migration] Running upgrade 1968acfc09e3 -> 2e82aab8ef20, rename user table
INFO  [alembic.runtime.migration] Running upgrade 2e82aab8ef20 -> 211e584da130, add TI state index
INFO  [alembic.runtime.migration] Running upgrade 211e584da130 -> 64de9cddf6c9, add task fails journal table
INFO  [alembic.runtime.migration] Running upgrade 64de9cddf6c9 -> f2ca10b85618, add dag_stats table
INFO  [alembic.runtime.migration] Running upgrade f2ca10b85618 -> 4addfa1236f1, Add fractional seconds to mysql tables
INFO  [alembic.runtime.migration] Running upgrade 4addfa1236f1 -> 8504051e801b, xcom dag task indices
INFO  [alembic.runtime.migration] Running upgrade 8504051e801b -> 5e7d17757c7a, add pid field to TaskInstance
INFO  [alembic.runtime.migration] Running upgrade 5e7d17757c7a -> 127d2bf2dfa7, Add dag_id/state index on dag_run table
INFO  [alembic.runtime.migration] Running upgrade 127d2bf2dfa7 -> cc1e65623dc7, add max tries column to task instance
usage: airflow [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: airflow --help [cmd1 cmd2 ...]
   or: airflow --help-commands
   or: airflow cmd --help

error: invalid command 'db'
```

**Comment by potiuk** [comment]
Maybe for some reason you installed Airlfow 1.10 over Airflow 2? For example if you try to install backport-providers, they are Airflow 1.10-only and when you install them from requirements, they will DOWNGRADE your airflow to 1.10....

**Reply by ziplokk1** [reply]
I'm running airflow 2.0.2 although your comment got me curious and all my provider versions are 1.0.x. I'm wondering if there's some compatibility issues. I'm going to update the providers and see if that remedies the issue.

```
❯ airflow version
2.0.2
❯ pip freeze | grep apache-airflow-providers
apache-airflow-providers-celery==1.0.1
apache-airflow-providers-ftp==1.0.0
apache-airflow-providers-http==1.0.0
apache-airflow-providers-imap==1.0.1
apache-airflow-providers-microsoft-mssql==1.0.0
apache-airflow-providers-mongo==1.0.0
apache-airflow-providers-postgres==1.0.1
apache-airflow-providers-redis==1.0.1
apache-airflow-providers-sftp==1.0.0
apache-airflow-providers-sqlite==1.0.1
apache-airflow-providers-ssh==1.0.0
```

**Comment by potiuk** [comment]
I think I saw it before. I can't remember the details, but I am quite sure that  one of your dags likely has some code that it should not suppose to run. I saw sometimes when people used to run their dags to test their DAGs as python code and added some code at the end of DAG that should never be run when DAGs are parsed by Airflow (this is what happens - your upgrade is actually parsing some DAGs).

What helped in that case was putting the code under:

```
if __name__ == '__main__':
     your test code here
```

**Reply by ziplokk1** [reply]
I'm curious why the upgrade parses the dags in the first place? Shouldn't the database migration be independent of the dags?

**Reply by potiuk** [reply]
Because of this line:

https://github.com/apache/airflow/blob/7e7fa620ccf23f804536799d6df66974e7b4f1a8/airflow/migrations/versions/cc1e65623dc7_add_max_tries_column_to_task_instance.py#L73

This is an old migration that actually create DagBAG and parses all DAG files in order to set "max_tries"  parameter value. This one appaently needs to instantiate DAG and all tasks because the "max_retries" value should be set to "retries" parameter of the task which can only be known if the DAG is parsed and task is instantiated.

It was long before SerializedDags table that could actually do that by reading serialized form from the DB. and it is pretty unfortunate that we have it for that reason only.

This is in the case when you already had some Task Instances in the DB  that needs that processing. But maybe your case is different when there is "clean" database the TI is empty (even if you have some DAGs and Tasks) - so we could skip the whole import if we find out that we have 0 Task Instances:

```
      session = sessionmaker(bind=connection)
      dagbag = DagBag(settings.DAGS_FOLDER)
```
 
add three lines:

```
      session = sessionmaker(bind=connection)
      if not bool(session.query(TaskInstance).first()):
          session.commit()
          return
      dagbag = DagBag(settings.DAGS_FOLDER)
```

Maybe you could try it and see if it works in your case ?

**Reply by potiuk** [reply]
Added a PR here: https://github.com/apache/airflow/pull/18450 in case this will work for you @ziplokk1

**Reply by potiuk** [reply]
I went away and merged it anyway

**Reply by ziplokk1** [reply]
Wow, thanks for the quick update @potiuk !!!

Your suggestion did fix the issue with that migration, though I'm getting the same error now at the end of the migrations and cant seem to pinpoint where it's coming from.

```
❯ airflow db init
DB: postgresql+psycopg2://airflow:***@localhost:5432/airflow
[2021-09-23 09:52:21,114] {db.py:684} INFO - Creating tables
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> e3a246e0dc1, current schema
INFO  [alembic.runtime.migration] Running upgrade e3a246e0dc1 -> 1507a7289a2f, create is_encrypted
INFO  [alembic.runtime.migration] Running upgrade 1507a7289a2f -> 13eb55f81627, maintain history for compatibility with earlier migrations
INFO  [alembic.runtime.migration] Running upgrade 13eb55f81627 -> 338e90f54d61, More logging into task_instance
INFO  [alembic.runtime.migration] Running upgrade 338e90f54d61 -> 52d714495f0, job_id indices
INFO  [alembic.runtime.migration] Running upgrade 52d714495f0 -> 502898887f84, Adding extra to Log
INFO  [alembic.runtime.migration] Running upgrade 502898887f84 -> 1b38cef5b76e, add dagrun
INFO  [alembic.runtime.migration] Running upgrade 1b38cef5b76e -> 2e541a1dcfed, task_duration
INFO  [alembic.runtime.migration] Running upgrade 2e541a1dcfed -> 40e67319e3a9, dagrun_config
INFO  [alembic.runtime.migration] Running upgrade 40e67319e3a9 -> 561833c1c74b, add password column to user
INFO  [alembic.runtime.migration] Running upgrade 561833c1c74b -> 4446e08588, dagrun start end
INFO  [alembic.runtime.migration] Running upgrade 4446e08588 -> bbc73705a13e, Add notification_sent column to sla_miss
INFO  [alembic.runtime.migration] Running upgrade bbc73705a13e -> bba5a7cfc896, Add a column to track the encryption state of the 'Extra' field in connection
INFO  [alembic.runtime.migration] Running upgrade bba5a7cfc896 -> 1968acfc09e3, add is_encrypted column to variable table
INFO  [alembic.runtime.migration] Running upgrade 1968acfc09e3 -> 2e82aab8ef20, rename user table
INFO  [alembic.runtime.migration] Running upgrade 2e82aab8ef20 -> 211e584da130, add TI state index
INFO  [alembic.runtime.migration] Running upgrade 211e584da130 -> 64de9cddf6c9, add task fails journal table
INFO  [alembic.runtime.migration] Running upgrade 64de9cddf6c9 -> f2ca10b85618, add dag_stats table
INFO  [alembic.runtime.migration] Running upgrade f2ca10b85618 -> 4addfa1236f1, Add fractional seconds to mysql tables
INFO  [alembic.runtime.migration] Running upgrade 4addfa1236f1 -> 8504051e801b, xcom dag task indices
INFO  [alembic.runtime.migration] Running upgrade 8504051e801b -> 5e7d17757c7a, add pid field to TaskInstance
INFO  [alembic.runtime.migration] Running upgrade 5e7d17757c7a -> 127d2bf2dfa7, Add dag_id/state index on dag_run table
INFO  [alembic.runtime.migration] Running upgrade 127d2bf2dfa7 -> cc1e65623dc7, add max tries column to task instance
INFO  [alembic.runtime.migration] Running upgrade cc1e65623dc7 -> bdaa763e6c56, Make xcom value column a large binary
INFO  [alembic.runtime.migration] Running upgrade bdaa763e6c56 -> 947454bf1dff, add ti job_id index
INFO  [alembic.runtime.migration] Running upgrade 947454bf1dff -> d2ae31099d61, Increase text size for MySQL (not relevant for other DBs' text types)
INFO  [alembic.runtime.migration] Running upgrade d2ae31099d61 -> 0e2a74e0fc9f, Add time zone awareness
INFO  [alembic.runtime.migration] Running upgrade d2ae31099d61 -> 33ae817a1ff4, kubernetes_resource_checkpointing
INFO  [alembic.runtime.migration] Running upgrade 33ae817a1ff4 -> 27c6a30d7c24, kubernetes_resource_checkpointing
INFO  [alembic.runtime.migration] Running upgrade 27c6a30d7c24 -> 86770d1215c0, add kubernetes scheduler uniqueness
INFO  [alembic.runtime.migration] Running upgrade 86770d1215c0, 0e2a74e0fc9f -> 05f30312d566, merge heads
INFO  [alembic.runtime.migration] Running upgrade 05f30312d566 -> f23433877c24, fix mysql not null constraint
INFO  [alembic.runtime.migration] Running upgrade f23433877c24 -> 856955da8476, fix sqlite foreign key
INFO  [alembic.runtime.migration] Running upgrade 856955da8476 -> 9635ae0956e7, index-faskfail
INFO  [alembic.runtime.migration] Running upgrade 9635ae0956e7 -> dd25f486b8ea, add idx_log_dag
INFO  [alembic.runtime.migration] Running upgrade dd25f486b8ea -> bf00311e1990, add index to taskinstance
INFO  [alembic.runtime.migration] Running upgrade 9635ae0956e7 -> 0a2a5b66e19d, add task_reschedule table
INFO  [alembic.runtime.migration] Running upgrade 0a2a5b66e19d, bf00311e1990 -> 03bc53e68815, merge_heads_2
INFO  [alembic.runtime.migration] Running upgrade 03bc53e68815 -> 41f5f12752f8, add superuser field
INFO  [alembic.runtime.migration] Running upgrade 41f5f12752f8 -> c8ffec048a3b, add fields to dag
INFO  [alembic.runtime.migration] Running upgrade c8ffec048a3b -> dd4ecb8fbee3, Add schedule interval to dag
INFO  [alembic.runtime.migration] Running upgrade dd4ecb8fbee3 -> 939bb1e647c8, task reschedule fk on cascade delete
INFO  [alembic.runtime.migration] Running upgrade 939bb1e647c8 -> 6e96a59344a4, Make TaskInstance.pool not nullable
INFO  [alembic.runtime.migration] Running upgrade 6e96a59344a4 -> d38e04c12aa2, add serialized_dag table
INFO  [alembic.runtime.migration] Running upgrade d38e04c12aa2 -> b3b105409875, add root_dag_id to DAG
INFO  [alembic.runtime.migration] Running upgrade 6e96a59344a4 -> 74effc47d867, change datetime to datetime2(6) on MSSQL tables
INFO  [alembic.runtime.migration] Running upgrade 939bb1e647c8 -> 004c1210f153, increase queue name size limit
INFO  [alembic.runtime.migration] Running upgrade c8ffec048a3b -> a56c9515abdc, Remove dag_stat table
INFO  [alembic.runtime.migration] Running upgrade a56c9515abdc, 004c1210f153, 74effc47d867, b3b105409875 -> 08364691d074, Merge the four heads back together
INFO  [alembic.runtime.migration] Running upgrade 08364691d074 -> fe461863935f, increase_length_for_connection_password
INFO  [alembic.runtime.migration] Running upgrade fe461863935f -> 7939bcff74ba, Add DagTags table
INFO  [alembic.runtime.migration] Running upgrade 7939bcff74ba -> a4c2fd67d16b, add pool_slots field to task_instance
INFO  [alembic.runtime.migration] Running upgrade a4c2fd67d16b -> 852ae6c715af, Add RenderedTaskInstanceFields table
INFO  [alembic.runtime.migration] Running upgrade 852ae6c715af -> 952da73b5eff, add dag_code table
INFO  [alembic.runtime.migration] Running upgrade 952da73b5eff -> a66efa278eea, Add Precision to execution_date in RenderedTaskInstanceFields table
INFO  [alembic.runtime.migration] Running upgrade a66efa278eea -> da3f683c3a5a, Add dag_hash Column to serialized_dag table
INFO  [alembic.runtime.migration] Running upgrade da3f683c3a5a -> 92c57b58940d, Create FAB Tables
INFO  [alembic.runtime.migration] Running upgrade 92c57b58940d -> 03afc6b6f902, Increase length of FAB ab_view_menu.name column
INFO  [alembic.runtime.migration] Running upgrade 03afc6b6f902 -> cf5dc11e79ad, drop_user_and_chart
INFO  [alembic.runtime.migration] Running upgrade cf5dc11e79ad -> bbf4a7ad0465, Remove id column from xcom
INFO  [alembic.runtime.migration] Running upgrade bbf4a7ad0465 -> b25a55525161, Increase length of pool name
INFO  [alembic.runtime.migration] Running upgrade b25a55525161 -> 3c20cacc0044, Add DagRun run_type
INFO  [alembic.runtime.migration] Running upgrade 3c20cacc0044 -> 8f966b9c467a, Set conn_type as non-nullable
INFO  [alembic.runtime.migration] Running upgrade 8f966b9c467a -> 8d48763f6d53, add unique constraint to conn_id
INFO  [alembic.runtime.migration] Running upgrade 8d48763f6d53 -> e38be357a868, Add sensor_instance table
INFO  [alembic.runtime.migration] Running upgrade e38be357a868 -> b247b1e3d1ed, Add queued by Job ID to TI
INFO  [alembic.runtime.migration] Running upgrade b247b1e3d1ed -> e1a11ece99cc, Add external executor ID to TI
INFO  [alembic.runtime.migration] Running upgrade e1a11ece99cc -> bef4f3d11e8b, Drop KubeResourceVersion and KubeWorkerId
INFO  [alembic.runtime.migration] Running upgrade bef4f3d11e8b -> 98271e7606e2, Add scheduling_decision to DagRun and DAG
INFO  [alembic.runtime.migration] Running upgrade 98271e7606e2 -> 52d53670a240, fix_mssql_exec_date_rendered_task_instance_fields_for_MSSQL
INFO  [alembic.runtime.migration] Running upgrade 52d53670a240 -> 364159666cbd, Add creating_job_id to DagRun table
INFO  [alembic.runtime.migration] Running upgrade 364159666cbd -> 45ba3f1493b9, add-k8s-yaml-to-rendered-templates
INFO  [alembic.runtime.migration] Running upgrade 45ba3f1493b9 -> 849da589634d, Prefix DAG permissions.
INFO  [alembic.runtime.migration] Running upgrade 849da589634d -> 2c6edca13270, Resource based permissions.
[2021-09-23 09:52:24,835] {manager.py:727} WARNING - No user yet created, use flask fab command to do it.
INFO  [alembic.runtime.migration] Running upgrade 2c6edca13270 -> 61ec73d9401f, Add description field to connection
INFO  [alembic.runtime.migration] Running upgrade 61ec73d9401f -> 64a7d6477aae, fix description field in connection to be text
INFO  [alembic.runtime.migration] Running upgrade 64a7d6477aae -> e959f08ac86c, Change field in DagCode to MEDIUMTEXT for MySql
INFO  [alembic.runtime.migration] Running upgrade e959f08ac86c -> 82b7c48c147f, Remove can_read permission on config resource for User and Viewer role
[2021-09-23 09:52:28,607] {manager.py:727} WARNING - No user yet created, use flask fab command to do it.
INFO  [alembic.runtime.migration] Running upgrade 82b7c48c147f -> 449b4072c2da, Increase size of connection.extra field to handle multiple RSA keys
INFO  [alembic.runtime.migration] Running upgrade 449b4072c2da -> 8646922c8a04, Change default pool_slots to 1
INFO  [alembic.runtime.migration] Running upgrade 8646922c8a04 -> 2e42bb497a22, rename last_scheduler_run column
INFO  [airflow.models.dagbag.DagBag] Filling up the DagBag from /Users/msanders/projects/sma
usage: airflow [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: airflow --help [cmd1 cmd2 ...]
   or: airflow --help-commands
   or: airflow cmd --help

error: invalid command 'db'
```

Running check-migrations seems to exit silently, albeit successfully.

```
❯ airflow db check-migrations
[2021-09-23 09:53:03,268] {migration.py:163} INFO - Context impl PostgresqlImpl.
[2021-09-23 09:53:03,268] {migration.py:166} INFO - Will assume transactional DDL.
❯ echo "$?"
0
```

But it looks like all the migrations were successfully applied.

```
❯ cd venv/lib/python3.9/site-packages/airflow
❯ alembic heads
2e42bb497a22 (head)
❯ alembic current
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
2e42bb497a22 (head)
```

But these are all problems for another time. Thank you so much for your detailed explanations and quick fix!

**Reply by potiuk** [reply]
Yeah. I think airflow db init performs Dag importing after finishing migration I believe.

I advise you to do bisecting. Remove half of your DAGs and run migration, Then either take those you have (if failed) or the other half (if not) and continue. You will quickly find the affected DAG this way and will be able to narrow down the search.

**Reply by potiuk** [reply]
I'd love to find out what it is and maybe find a solution to warn better if others will have similar problem - I already saw it once so it can happen again

**Reply by ziplokk1** [reply]
I've solved the issue. It was multiple problems.

Putting a `setup.py` which utilizes `setuptools.find_packages` in the root of the `AIRFLOW__CORE__DAGS_FOLDER` causes the silent failure.

After removing the `setup.py` file the migration still failed, but this time it was because my `AIRFLOW__CORE__DAGS_FOLDER` was my project root (due to dependencies in the dags) which also contained my virtualenv. This caused the following error - which I'm assuming is because it's attempting to zip the contents of the dag bag into a zip which is itself located in a directory which is being zipped; but I'm just taking a guess. Here's the traceback:

```
ERROR [airflow.models.dagbag.DagBag] Bad magic number for central directory
Traceback (most recent call last):
  File "/Users/msanders/projects/sma/venv/lib/python3.9/site-packages/airflow/models/dagbag.py", line 454, in collect_dags
    found_dags = self.process_file(filepath, only_if_updated=only_if_updated, safe_mode=safe_mode)
  File "/Users/msanders/projects/sma/venv/lib/python3.9/site-packages/airflow/models/dagbag.py", line 264, in process_file
    mods = self._load_modules_from_zip(filepath, safe_mode)
  File "/Users/msanders/projects/sma/venv/lib/python3.9/site-packages/airflow/models/dagbag.py", line 308, in _load_modules_from_zip
    current_zip_file = zipfile.ZipFile(filepath)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/zipfile.py", line 1257, in __init__
    self._RealGetContents()
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/zipfile.py", line 1352, in _RealGetContents
    raise BadZipFile("Bad magic number for central directory")
zipfile.BadZipFile: Bad magic number for central directory
```

In conclusion:

1) Don't put a `setup.py` which uses `setuptools.find_packages()` in your `AIRFLOW__CORE__DAGS_FOLDER`.
2) Don't put airflow's virtualenv in your `AIRFLOW__CORE__DAGS_FOLDER`.
